docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+       oe_get_evidence(&format, 0, NULL, 0, opt_paramsY, sizeof(opt_paramsY), &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   uint8_t* opt_params;
+   size_t opt_params_size;
+   oe_get_evidence_parameters(&format, &opt_params, &opt_params_size);
+
+   oe_get_evidence(&format, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+4. SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. The local/remote attestation concept is SGX specific. To prevent confusion at evidence generation, the OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION and OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION flags were deprecated.
+
+```C
+   // Old
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+   oe_get_evidence(&format_remote_only, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+
+   // New
+   oe_get_evidence(&format_sgx_local, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx_remote, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_remote_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+User Experience
+---------------
+The user experience is improved by the changes. These are a couple of possible attestation scenarios:
+
+Scenario 1 - Attester selects from a list of evidence formats that can be verified:
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
+    2. Builds the format IDs of the registered verifiers sorted by preference.
+    3. Sends the format IDs to the attester application. 
+- Attester application
+    1. Upon initialization, queries available attesters to the application (oe_sgx_get_attesters) and registers them (oe_register_attester).
+    2. Receives format IDs from the verifier application.
+    3. Selects a format ID from the received format IDs (oe_select_attester_format).
+    4. Sends the format ID to the verifier application.
+- Verifier application
+    1. Gets evidence parameters for the received format ID (oe_get_evidence_parameters).
```

Comment:
What does that mean?  I don't follow.

Reply:
Some evidence generation needs info from verifier as as input. For example, SGX local report. This step allows the verifier plugin to output parameters needed to generate evidence. For other evidence formats, this can be empty.

---
In reply to: [393256947](https://github.com/openenclave/openenclave/pull/2621#discussion_r393256947) [](ancestors = 393256947)

Reply:
Just elaborate in the text more, the point is valid (thanks for the explanation in github response).  If I understand correctly, the nonce discussed in the rats-architecture and EAT documents would be an "evidence parameter" here, and that makes perfect sense.  You might use the nonce example in the text here, since the term "evidence parameter" appears nowhere in external documents about attestation architecture.

Reply:
A few sentences appended to explain the purpose of 'evidence parameters'.

---
In reply to: [394655824](https://github.com/openenclave/openenclave/pull/2621#discussion_r394655824) [](ancestors = 394655824)